pkg/flashdb: bump version and improve test

MichelRottleuthner opened the pull request:
### Contribution description
Bumps the FlashDB package to current master and improves the test a bit.
The current version in RIOT has a bug in the file-based storage implementation. The bug breaks the libc based file access and therefore breaks the file-based mode on all boards except native, which uses the posix mode file access.

During testing I noticed the test apps are not easy to interpret on the first look. They also don't actually work on real hardware. The test in master swallows most errors silently and in most cases doesn't check return codes. Some values were not persisted across reboots (e.g., the boot_counter) as they got lost somewhere in the cache.

I added some output to the test that makes it easier to see if something went wrong. Explicitly calling the de-init functions of FlashDB in the test application fixes the issue with the broken persistence.

### Testing procedure
To test on a board you should make sure the persistent storage is clean, for example:
`make BOARD=adafruit-feather-nrf52840-sense -C tests/sys/vfs_default all flash term`
Then format through the RIOT shell:
```
> vfs umount /nvm0
> format /nvm0
```
Run the flashdb_vfs test:
`USEMODULE=test_utils_interactive_sync make BOARD=adafruit-feather-nrf52840-sense -C tests/pkg/flashdb_vfs all flash term`
```
# START
# main(): This is RIOT! (Version: 2026.01-devel-158-g522b2-flashdb_test_fixes)
# mkdir '/nvm0/fdb_kvdb1' [OK]
# ==================== kvdb_basic_sample ====================
# get the 'boot_count' value is 0 [OK]
# set the 'boot_count' value to 1 [OK]
# get the 'boot_count' value is 1 [OK]
# ===========================================================
# ================= kvdb_type_string_sample =================
# create the 'temp' string KV, value is: 36C [OK]
# get the 'temp' value is: 36C [OK]
# set 'temp' value to 38C [OK]
# delete the 'temp' finish [OK]
# ===========================================================
# ================== kvdb_type_blob_sample ==================
# create the 'temp' blob KV, value is: 36 [OK]
# get the 'temp' value is: 36 [OK]
# set 'temp' value to 38 [OK]
# delete the 'temp' finish [OK]
# ===========================================================
# kvdb deinit [OK]
# mkdir '/nvm0/fdb_tsdb1' [OK]
# ======================= tsdb_sample =======================
# append the new status.temp (36) and status.humi (85) [OK]
# append the new status.temp (38) and status.humi (90) [OK]
# [query_cb] queried a TSL: time: 11, temp: 36, humi: 85 [OK]
# [query_cb] queried a TSL: time: 12, temp: 38, humi: 90 [OK]
# [query_by_time_cb] queried a TSL: time: 11, temp: 36, humi: 85 [OK]
# [query_by_time_cb] queried a TSL: time: 12, temp: 38, humi: 90 [OK]
# query count is: 2
# set the TSL (time 11) status from 2 to 3 [OK]
# set the TSL (time 12) status from 2 to 3 [OK]
# ===========================================================
# tsdb deinit [OK]
```
***smash that reset button***
```
# START
# main(): This is RIOT! (Version: 2026.01-devel-158-g522b2-flashdb_test_fixes)
# mkdir '/nvm0/fdb_kvdb1' [OK]
# ==================== kvdb_basic_sample ====================
# get the 'boot_count' value is 1 [OK]
# set the 'boot_count' value to 2 [OK]
# get the 'boot_count' value is 2 [OK]
# ===========================================================
# ================= kvdb_type_string_sample =================
# create the 'temp' string KV, value is: 36C [OK]
# get the 'temp' value is: 36C [OK]
# set 'temp' value to 38C [OK]
# delete the 'temp' finish [OK]
# ===========================================================
# ================== kvdb_type_blob_sample ==================
# create the 'temp' blob KV, value is: 36 [OK]
# get the 'temp' value is: 36 [OK]
# set 'temp' value to 38 [OK]
# delete the 'temp' finish [OK]
# ===========================================================
# kvdb deinit [OK]
# mkdir '/nvm0/fdb_tsdb1' [OK]
# ======================= tsdb_sample =======================
# append the new status.temp (36) and status.humi (85) [FAILED]
# append the new status.temp (38) and status.humi (90) [FAILED]
# [query_cb] queried a TSL: time: 11, temp: 36, humi: 85 [OK]
# [query_cb] queried a TSL: time: 12, temp: 38, humi: 90 [OK]
# [query_by_time_cb] queried a TSL: time: 11, temp: 36, humi: 85 [OK]
# [query_by_time_cb] queried a TSL: time: 12, temp: 38, humi: 90 [OK]
# query count is: 0
# set the TSL (time 11) status from 3 to 3 [OK]
# set the TSL (time 12) status from 3 to 3 [OK]
# ===========================================================
# tsdb deinit [OK]
```
Same for `tests/pkg/flashdb_mtd`. Both applications should also work on native (for me they did).

Note that the failure with the time series append operations happens due to the dummy timer implementation of the test. That also happened before but wasn't visible. 

### Issues/PRs references

The package repo actually [some more tests](https://github.com/armink/FlashDB/blob/master/tests/fdb_kvdb_tc.c). We should probably integrate those in the future instead of the simplified examples we currently have as our tests.
<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

MichelRottleuthner committed:
tests/pkg/flashdb_vfs: improve test output and increase stacksize

The stacksize was not large enough to run this on an embedded target with flash based mtd and littlefs2.
The test now checks return values for expected values and reports issues to the user.

MichelRottleuthner committed:
tests/pkg/flashdb_vfs: improve kvdb_basic_sample test and output

MichelRottleuthner committed:
tests/pkg/flashdb_vfs: improve kvdb_type_string_sample test and output

MichelRottleuthner committed:
tests/pkg/flashdb_vfs: improve kvdb_type_blob_sample test and output

MichelRottleuthner committed:
tests/pkg/flashdb_vfs: improve tsdb_sample test and output

MichelRottleuthner committed:
tests/pkg/flashdb_vfs: deinit kvdb and tsdb to close files

Without this change, some changes to the persistent storage (e.g., with
littlefs2 on mtd_spi_nor) are not actually written back to the 'disk'.
The boot_count in the example was therefore not persistent across reboots.
Deinitializing solves this by properly closing all open files.

leandrolanzieri reviewed (commented):
Nice! Thanks for this.

Do you mind applying the static check suggestions now that we're touching the file anyway? I'll run this on hardware tomorrow.

leandrolanzieri commented on tests/pkg/flashdb_vfs/kvdb_type_blob_sample.c:
```diff
@@ -17,45 +17,42 @@
 #define FDB_LOG_TAG "[sample][kvdb][blob]"
 
+extern const char *resstr(bool condition);
+
```
Small typo in message of 7dd31e94e36486d5146b2991737fc987c8eb234b: `sampl` -> `sample`

leandrolanzieri commented on tests/pkg/flashdb_vfs/main.c:
```diff
@@ -108,7 +108,7 @@ int main(void)
         /* create database directory */
         int res = vfs_mkdir(KVDB_DIR, 0777);
-        printf("mkdir '%s' %s", KVDB_DIR, resstr(res == 0 || res == -EEXIST));
+        printf("mkdir '%s' %s\n", KVDB_DIR, resstr(res == 0 || res == -EEXIST));
```
Here and below: probably belong to 0a1b4a644b2c8103a5d9ab6019c220a00b802ef7

MichelRottleuthner committed:
pkg/flashdb: bump version

This bumps the package version to the current master as the most recent release
has a bug in the LIBC-flavour of the file-based storage backend. Without this
fix the file-based sotrage is therefore not usable on any target other than
native.

MichelRottleuthner committed:
tests/pkg/flashdb_vfs: apply static-tests suggestions

mguetschow commented on tests/pkg/flashdb_vfs/kvdb_basic_sample.c:
```diff
@@ -26,21 +29,26 @@ void kvdb_basic_sample(fdb_kvdb_t kvdb)
     { /* GET the KV value */
         /* get the "boot_count" KV value */
-        fdb_kv_get_blob(kvdb, "boot_count", fdb_blob_make(&blob, &boot_count, sizeof(boot_count)));
-        /* the blob.saved.len is more than 0 when get the value successful */
```
I was under the impression most (all?) the tests were taken from their docs: https://armink.github.io/FlashDB/#/sample-kvdb-basic

I haven't done so yet, but maybe you could double-check if they have changed in the meantime. Nevertheless, it definitely makes sense to catch return values.

Will try to have a closer look at it today or beginning of next week.

crasbe commented on tests/pkg/flashdb_vfs/tsdb_sample.c:
```diff
@@ -102,8 +106,9 @@ static bool query_by_time_cb(fdb_tsl_t tsl, void *arg)
-    fdb_blob_read((fdb_db_t) db, fdb_tsl_to_blob(tsl, fdb_blob_make(&blob, &status, sizeof(status))));
-    printf("[query_by_time_cb] queried a TSL: time: %"PRId32", temp: %d, humi: %d\n", tsl->time, status.temp, status.humi);
+    size_t len = fdb_blob_read((fdb_db_t) db, fdb_tsl_to_blob(tsl, fdb_blob_make(&blob, &status, sizeof(status))));
+    printf("[query_by_time_cb] queried a TSL: time: %"PRId32", temp: %d, humi: %d %s\n", tsl->time, status.temp, status.humi,
```
Same.

MichelRottleuthner committed:
fixup! tests/pkg/flashdb_vfs: improve kvdb_basic_sample test and output

MichelRottleuthner committed:
fixup! tests/pkg/flashdb_vfs: improve kvdb_type_blob_sample test and output

MichelRottleuthner committed:
fixup! tests/pkg/flashdb_vfs: improve tsdb_sample test and output

MichelRottleuthner commented on tests/pkg/flashdb_vfs/kvdb_basic_sample.c:
```diff
@@ -27,21 +29,26 @@ void kvdb_basic_sample(fdb_kvdb_t kvdb)
         /* change the "boot_count" KV's value */
-        fdb_kv_set_blob(kvdb, "boot_count", fdb_blob_make(&blob, &boot_count, sizeof(boot_count)));
-        printf("set the 'boot_count' value to %d\n", boot_count);
+        int res = fdb_kv_set_blob(kvdb, "boot_count", fdb_blob_make(&blob, &boot_count, sizeof(boot_count)));
```
Yeah. Probably was not caught because many of the test lines were also too long before. I wanted to keep the changes small. But, well, while we are at it... Made all the lines <100 chars now.

mguetschow commented on tests/pkg/flashdb_vfs/kvdb_basic_sample.c:
```diff
@@ -26,21 +29,29 @@ void kvdb_basic_sample(fdb_kvdb_t kvdb)
+    { /* GET the KV value */
+        int expected = boot_count;
+        /* get the "boot_count" KV value */
+        int res = fdb_kv_get_blob(kvdb, "boot_count",
```
(could also be its own function as this is basically duplicated from above)

mguetschow reviewed (commented):
Nice! Smaller nit's below. Now that we have the [OK] values, it wouldn't be to hard to add a Python test for it (maybe in a follow-up PR, though) :)

AnnsAnns commented:
@MichelRottleuthner :pleading_face: any chance for the final nitpicks to be fixed?
